File: Advanced Algorithms/rsa.py
import random
from math import gcd

# ------------------ HELPER FUNCTIONS ------------------

# Modular inverse using Extended Euclidean Algorithm
def mod_inverse(e, phi):
    def extended_gcd(a, b):
        if b == 0:
            return a, 1, 0
        g, x1, y1 = extended_gcd(b, a % b)
        x = y1
        y = x1 - (a // b) * y1
        return g, x, y

    g, x, _ = extended_gcd(e, phi)
    if g != 1:
        raise Exception("Modular inverse does not exist")
    return x % phi


# Check for prime (simple version for learning)
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True


# Generate a random prime number (small for demo)
def generate_prime(start=100, end=300):
    while True:
        num = random.randint(start, end)
        if is_prime(num):
            return num


# ------------------ RSA KEY GENERATION ------------------

def generate_keys():
    p = generate_prime()
    q = generate_prime()

    while q == p:
        q = generate_prime()

    n = p * q
    phi = (p - 1) * (q - 1)

    # Choose e such that gcd(e, phi) = 1
    e = random.randrange(2, phi)
    while gcd(e, phi) != 1:
        e = random.randrange(2, phi)

    # Compute d
    d = mod_inverse(e, phi)

    return (e, n), (d, n)


# ------------------ ENCRYPTION ------------------

def encrypt(plaintext, public_key):
    e, n = public_key
    return [pow(ord(char), e, n) for char in plaintext]


# ------------------ DECRYPTION ------------------

def decrypt(ciphertext, private_key):
    d, n = private_key
    return ''.join([chr(pow(char, d, n)) for char in ciphertext])


# ------------------ MAIN ------------------

if __name__ == "__main__":
    public_key, private_key = generate_keys()

    print("Public Key (e, n):", public_key)
    print("Private Key (d, n):", private_key)

    message = "HELLO RSA"

    # Encrypt
    encrypted_msg = encrypt(message, public_key)
    print("Encrypted:", encrypted_msg)

    # Decrypt
    decrypted_msg = decrypt(encrypted_msg, private_key)
    print("Decrypted:", decrypted_msg)

File: Advanced Algorithms/rsa.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function to compute gcd
int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Modular exponentiation (square and multiply)
long long mod_exp(long long base, long long exp, long long mod) {
    long long result = 1;
    base = base % mod;

    while (exp > 0) {
        if (exp % 2 == 1)
            result = (result * base) % mod;

        base = (base * base) % mod;
        exp = exp / 2;
    }
    return result;
}

// Extended Euclidean Algorithm for modular inverse
long long mod_inverse(long long e, long long phi) {
    long long t = 0, newt = 1;
    long long r = phi, newr = e;

    while (newr != 0) {
        long long q = r / newr;

        long long temp = newt;
        newt = t - q * newt;
        t = temp;

        temp = newr;
        newr = r - q * newr;
        r = temp;
    }

    if (r > 1) {
        printf("Inverse does not exist\n");
        exit(1);
    }

    if (t < 0)
        t += phi;

    return t;
}

// Simple prime check
int is_prime(int n) {
    if (n < 2) return 0;
    for (int i = 2; i * i <= n; i++) {
        if (n % i == 0)
            return 0;
    }
    return 1;
}

// Generate small random prime (for demo)
int generate_prime() {
    while (1) {
        int num = rand() % 100 + 50; // range 50–149
        if (is_prime(num))
            return num;
    }
}

int main() {
    srand(time(NULL));

    int p = generate_prime();
    int q = generate_prime();

    while (q == p)
        q = generate_prime();

    long long n = (long long)p * q;
    long long phi = (long long)(p - 1) * (q - 1);

    // Choose e
    long long e = 2;
    while (e < phi) {
        if (gcd(e, phi) == 1)
            break;
        e++;
    }

    // Compute d
    long long d = mod_inverse(e, phi);

    printf("Public Key (e, n): (%lld, %lld)\n", e, n);
    printf("Private Key (d, n): (%lld, %lld)\n", d, n);

    char message[] = "HELLO";
    int len = sizeof(message) - 1;

    long long encrypted[100];

    // Encryption
    printf("Encrypted: ");
    for (int i = 0; i < len; i++) {
        encrypted[i] = mod_exp(message[i], e, n);
        printf("%lld ", encrypted[i]);
    }
    printf("\n");

    // Decryption
    printf("Decrypted: ");
    for (int i = 0; i < len; i++) {
        char decrypted_char = (char)mod_exp(encrypted[i], d, n);
        printf("%c", decrypted_char);
    }
    printf("\n");

    return 0;
}